byteps/server/server.cc

```diff
@@ -134,144 +174,191 @@ void BytePSServerEngineThread(int i) {
           if (debug_mode_ && (debug_key_ == key)) {
             std::lock_guard<std::mutex> lock(debug_mu_);
             LOG(INFO) << "stage: FIRST_WORKER_RECV \t"
-                      << "stored: " << DEBUG_PRINT_TENSOR_VALUE(stored->tensor) << "\t"
+                      << "stored: " << DEBUG_PRINT_TENSOR_VALUE(stored->tensor)
+                      << "\t"
                       << "recved: " << DEBUG_PRINT_TENSOR_VALUE(recved) << "\t"
                       << "len: " << len << "\t"
                       << "addr: " << DEBUG_PRINT_TENSOR_ADDRESS(recved);
           }
-          // zero copy
-          updates.merged.tensor = recved;
           updates.merged.tmp_sarray = req_data;
-        } else { // async mode, directly add to the buffer
-          CHECK_GE(bps_reducer_->sum((void *) stored->tensor,
-                                    (void *) recved,
-                                    len,
-                                    bps_reducer_->GetDataType(stored->dtype)), 0);
+          // copy
+          BytePSEngineMessage msg = {timestamp_++,   type,     key,
+                                     stored->tensor, recved,   stored->len,
+                                     COPY_FIRST,     req_data, req_meta};
+          engine_queues_[tid]->Push(msg);
+        } else {  // async mode, directly add to the buffer
+          CHECK_GE(bps_reducer_->sum((void*)stored->tensor, (void*)recved, len,
+                                     bps_reducer_->GetDataType(stored->dtype)),
+                   0);
         }
-      } else { // from other workers
+      } else {  // from other workers
         CHECK(sync_mode_);
-        CHECK(updates.merged.tensor);
+        // CHECK(updates.merged.tensor);
         if (debug_mode_ && (debug_key_ == key)) {
           std::lock_guard<std::mutex> lock(debug_mu_);
           LOG(INFO) << "stage: OTHER_WORKER_SUM \t"
-                    << "stored: " << DEBUG_PRINT_TENSOR_VALUE(stored->tensor) << "\t"
-                    << "merged: " << DEBUG_PRINT_TENSOR_VALUE(updates.merged.tensor) << "\t"
+                    << "stored: " << DEBUG_PRINT_TENSOR_VALUE(stored->tensor)
+                    << "\t"
                     << "recved: " << DEBUG_PRINT_TENSOR_VALUE(recved) << "\t"
                     << "len: " << len << "\t"
                     << "addr: " << DEBUG_PRINT_TENSOR_ADDRESS(recved);
         }
         if (is_engine_blocking_) {
-          CHECK_GE(bps_reducer_->sum((void *) updates.merged.tensor,
-                                    (void *) recved,
-                                    len,
-                                    bps_reducer_->GetDataType(updates.merged.dtype)), 0);
-        } else { // non-blocking
-          BytePSEngineMessage msg = {timestamp_++, type, key, updates.merged.tensor, recved, len, SUM_RECV, req_data, req_meta};
+          // TODO: decompress
+          CHECK_GE(bps_reducer_->sum(
+                       (void*)updates.merged.tensor, (void*)recved, len,
```

Comment:
`updates.merged.tensor` here is used before it's initialized. The `ALL_RECV` message for this key is pushed to the queue on L362, then `updates.merged.tensor` is assigned for the first time on L110. Am I missing anything here?

Reply:
You are right. It will be fixed in the future. Currently it only support nonblocking sync case.

Reply:
Not sure if I grok that. "nonblocking" and "sync" are two mutually exclusive modes, the branch here is the synchronous mode, no? The core dump I saw is caused by this line.

Reply:
I am not sure I get your point. You mean the server crashes at L327-L332? It is expected because blocking mode is not supported yet. 

BTW, nonblocking and sync are not contradictory. Non-blocking just means jobs are pushed to the backend working threads. Currently compression / decompression is done in `BytePSServerEngineThread`.

Reply:
Thanks for the explanation. `updates.merged.tensor` here is not a problem then.

> I am not sure I get your point. You mean the server crashes at L327-L332? It is expected because blocking mode is not supported yet.
> 
> BTW, nonblocking and sync are not contradictory. Non-blocking just means jobs are pushed to the backend working threads. Currently compression / decompression is done in `BytePSServerEngineThread`.